### DISCUSSION TITLE
Resource Optimization

### DISCUSSION THREAD

**Discussion by paulobreim** [sepinf-inc/IPED#1899] [Q&A]
I have a computer with good configuration where I am indexing a case with 35 cell phones and it has already been running for more than 100 hours and should still take another 100.
Looking at the task manager, I'm using 8% CPU and 30% memory. I have 3 SSDs (read, write and index) all with 0%. The biggest time spent on the IPED is AudioTranscription task at 97%, as expected.

I would like to know if I can start another case on this same machine.

**Comment by wladimirleite** [comment]
What audio transcription implementation are you using? (Vosk, Wav2Vec2 etc.)

**Reply by paulobreim** [reply]
I am using Vosk.

**Reply by wladimirleite** [reply]
As **AudioTranscriptionTask** is dominating the processing time, the expected behavior would be CPU usage close to 100%. 
So in theory, I wouldn't be a good idea to start another process.

But as you noticed, CPU usage is low, so I guess that something wrong is going on.
Can you check the processing log?

**Reply by paulobreim** [reply]
the log is large, but I didn't see anything exceptional. on the other hand, it has a lot of technical information, which is difficult to analyze without being involved in the development.
I'm attaching the log if you want to look.
[IPED-2023-09-22-19-41-30.zip](https://github.com/sepinf-inc/IPED/files/12729808/IPED-2023-09-22-19-41-30.zip)

Do you think you should stop and use another type of transcription?

**Reply by lfcnassif** [reply]
I suggest taking a screenshot of the processing UI to see what is the running Task on each worker (last table to the right) when the CPU and disks usage are both low. And also to take a thread dump using `jstack -l pid` if most workers are running the same Task (in the same low CPU and disks usage situation), maybe there is some kind of bottleneck the thread dump may bring to light.

**Reply by wladimirleite** [reply]
I took a quick look at the log, but couldn't find anything wrong/suspicious.
Vosk should work fine. I used it a lot of times without any kind of trouble, and it was relatively fast.
Now we are using RemoteWav2Vec2, mainly because of its much better transcription quality, but we have a dedicated internal server (with many high-end GPUs).
There is a option to use Wav2Vec2 locally, but if you don't have a good GPU (with a lot of memory), it will use the CPU and should be way too slow for most use cases.

One last thing to check, if you open the case, and search for "audio\:transcription:*", can you see a lot of audios (and their transcription)?

**Reply by paulobreim** [reply]
Well, the "audio:transcription:*" has 384,798 files, so it seems to be doing ok.
I'm using iped.exe and not the .jar, I don't know if it makes any difference and I don't even have any Java development installed, I think that's why I don't have jstack.
I didn't quite understand where to find this information that @lfcnassif  na explained.
But I printed some screens below for possible analysis.
It's really strange that the machine is so idle.
![Screenshot_20230926_192101](https://github.com/sepinf-inc/IPED/assets/381594/4c9a2480-85a7-4eb9-a76d-dacdee311d12)
![Screenshot_20230926_192026](https://github.com/sepinf-inc/IPED/assets/381594/40a46d38-44c3-40be-8fbd-8cc70ff28145)
![Screenshot_20230926_191949](https://github.com/sepinf-inc/IPED/assets/381594/d5a604f1-850a-44d4-8052-2c930cba054d)
![Screenshot_20230926_191827](https://github.com/sepinf-inc/IPED/assets/381594/b34342b3-04b0-434f-8e07-7dc53b44e565)
The total time elapsed is about 30 hours longer, because I had to restart with --continue.

**Reply by lfcnassif** [reply]
Hi @paulobreim, thanks. I think in this case the `jstack -l pid` output is very important to diagnose the bottleneck. You will have to install some JDK 11, or maybe there is a standalone jstack tool out there.

**Reply by wladimirleite** [reply]
@paulobreim, I made a quick test using Vosk with IPED 4.1.4, processing a folder with a few thousand WhatsApp audios. AudioTranscriptTask used ~90% of the processing time and CPU usage was very high, as expected. I have 48 logical processors here.

**Reply by paulobreim** [reply]
I started installing some Java stuff and IPED died, probably Java had some anomaly because when I restarted IPED, it started to take up the entire CPU and the number of files processed in the audio transcription task accelerated a lot.
I'll keep an eye out to see if it happens again.

![Screenshot_20230926_234946](https://github.com/sepinf-inc/IPED/assets/381594/a239a8a0-7a8a-47bb-9723-4e3105f5d4cf)
![Screenshot_20230926_235808](https://github.com/sepinf-inc/IPED/assets/381594/d99e38be-5e3e-4a49-b314-5a8bc74bf4c7)

Tanks for help

**Reply by wladimirleite** [reply]
Hi @paulobreim! How is the processing going?
Thinking about this later, one hypothesis that I considered was that somehow Vosk entered in a loop or some race condition. It uses native code, so it may have shared resources outside the JVM. I never saw a similar situation, but in theory it is possible. 
Not sure, but I think that there is no timeout control in the call to Vosk audio transcription.
If that was the case, some of the IPED workers could have been stuck in the same audio file.
So if it happens again, besides running jstack as @lfcnassif suggested, it would be interesting to scroll the processing window to the right, so the file names become visible, and see if after a few seconds it is changing to another file or not. 

Not directly related to this, but when I made some visual changes in the processing window, I considered adding another narrow column, with the time spent so far in that item/task. So if some items are taking longer (or even stuck), that would be easier to spot.

**Reply by wladimirleite** [reply]
I think I managed to reproduce the situation described by @paulobreim, in which all workers are running AudioTranscriptTask, but CPU usage is low (and almost no I/O and no memory contention).

Processing a lot of OPUS audios (using Vosk), the process started using ~100% of CPU, which is the expected behavior.
After 40 minutes or so, the average CPU usage fell to ~50%.
Now, after 8 hours, it is around 20-25%.

There are no workers "frozen"/"trapped in a loop" as audio files keep changing after some time (I added the column that I mentioned before that makes it easier to spot how long each item has been in the current task).
So the conclusion is that some kind of internal lock/synchronized code is getting slower after processing a lot of audio files, and adding a timeout control won't help.
I will try to investigate a bit more, and see if I can find a solution (or at least a workaround).
It is not critical because it only happened when I processed a lot of audio files, so in most scenarios this "slow down" won't happen.

**10 minutes**
![image](https://github.com/sepinf-inc/IPED/assets/7217001/603420ff-40a2-4b43-b82f-612f8621f979)
 
**40 minutes**
![image](https://github.com/sepinf-inc/IPED/assets/7217001/e20f730f-f842-4c3b-8b03-89527e210312)

**8 hours**
![image](https://github.com/sepinf-inc/IPED/assets/7217001/4731f1e1-0d80-4692-a599-fd5531fd8820)

**Reply by lfcnassif** [reply]
That's really weird, thanks @tc-wleite for dedicating time to reproduce. If you could take a thread dump with jstack, it may help to investigate, if the issue is in the java code side...

**Reply by paulobreim** [reply]
Very good.
Exactly the same behavior here. It starts at 100% and decreases.
After 4 hours it is at 25% CPU.
Since it now takes about 1 hour to get to where it was, if I restart, I don't know whether to restart or let it continue. I think I'll leave it to see where this CPU goes.

**Comment by lfcnassif** [comment]
> Not sure, but I think that there is no timeout control in the call to Vosk audio transcription.

That's correct. Actually just the mplayer conversion has a timeout, the Vosk transcription itself no. When implementing, since Vosk receives raw WAV data, I thought it would be unlikely to enter in a loop inside Vosk, but @tc-wleite is right, it is possible in theory, and IPED multithreading might trigger race condition issues in Vosk, if they exist.

**Reply by paulobreim** [reply]
Hello gentlemen,
I just got back from a trip, I went to clone some things in the city of Cruz Alta (very far away).
After that, the processing that was being performed finished the ufdr quickly. This ufdr was around 39 GB.
Then it processed another 95GB without any problems.
Then it started processing a 130GB one, but it ended up giving an out of memory error. I restarted it twice and it kept giving the same e
[IPED-2023-09-27-19-14-09.log](https://github.com/sepinf-inc/IPED/files/12745330/IPED-2023-09-27-19-14-09.log)
rror.

I chose to remove this file in -d and continue instead of trying a -Xms30G -Xmx30G because every time I restart IPED, it takes about 45 minutes to get to the point it was in, basically with processing in TempFileTask. Shouldn't this be faster since it already knows which files were processed just by looking at the name? I understand that IPED opens the file because although the name is the same, it could be a different file. As I don't know which one was actually finished, I've been keeping them all in -d.
I looked at this file on Celebritte and the extraction process started at 4:14 pm one day and ended at 6:58 am the next day!
But then I'll run this file in isolation to see what happens.
Now there are 12 files left, let's see what happens.
After everything is finished, I intend to reprocess everything with another type of transcription

tks
paulo

**Reply by lfcnassif** [reply]
> Then it started processing a 130GB one, but it ended up giving an out of memory error. I restarted it twice and it kept giving the same error.

There are tips to workaround the OOM in the wiki:
https://github.com/sepinf-inc/IPED/wiki/Troubleshooting#outofmemoryerror

If none of them work, we would need the generated heap dump (*.hprof) to investigate...

**Reply by paulobreim** [reply]
Yes, IPED is smart and show this link at the end of log.
When this processing is finished, now analyzing a 137GB ufdr and there are 10 ufdr more, I will reprocess the ufdr that caused the error, alone, and then use these alternatives.
I'm enjoying these tests, it's a real case but at the same time it's excellent for testing.

**Reply by paulobreim** [reply]
@lfcnassif Is it possible for you to make a modification to IPED, to record it in the Log when a file is finished?
For example: indexing of file xxxx.ufdr finished.
Having this information, when the user have to restart a process, it would be easy to remove this file from -d.

**Reply by lfcnassif** [reply]
It is not possible today, some kind of artifacts, e.g. SQLite databases, are sent to a second processing queue, regardless of the evidence they came from. I mean, the tool first process all basic artifacts from evidence 1, then evidence 2, then 3. After that, the tool process complex artifacts from evidence 1, then 2, then 3. Finally, more complex artifacts are processed from each evidence. So, the tool doesn't process each evidence sequentially, but it processes artifacts sequentially based on "complexity".

**Reply by paulobreim** [reply]
Windows mysteries.
Two more times the CPU became low again.
I ran task manager and used the PID details option and there is the option "Analyze Wait Chain". It takes a few seconds and shows the screen below.
The problem is that right away it simply kills the application!! This is a Windows BUG, because I tested this string analysis in Firefox and it killed Firefox.
I tested it several times with different applications, every time there is a chain wait, it shows the chain but then the process dies, if there is no wait it gives a message that the application is executing normally.
I still haven't been able to install jstack.
![Screenshot_20230929_124734](https://github.com/sepinf-inc/IPED/assets/381594/d874f5d9-0bc0-446b-b902-92a625c86ae9)

**Reply by wladimirleite** [reply]
Not sure if jstack would help, as it seems to be some issue in native memory/code. As I was able to reproduce the same behavior, now it is "only" a matter of finding a solution.

**Reply by paulobreim** [reply]
But I found it very strange that the process died just because a query was made to see if there is a waiting chain, it could be a Windows bug, which causes the process to die, even because even a Firefox one dies.

Vosk is probably deadlocked because all 56 threads were on this task.

**Reply by paulobreim** [reply]
Now it started to go fast because the audio transcription of this ufdr was finished.
The CPU has risen to 80% and there are many DIEtask (what is this task?)
There are too a lot of QRCodeTask, but in my IPEDConfig is set to enableQRCode = false

**Reply by wladimirleite** [reply]
It is a task to detect explicit (nudity) images. There is a comment about that in IPEDConfig.txt. 
It usually doesn't consume a high percentage of processing time.

**Reply by paulobreim** [reply]
New situation:
Another ufdr gave out of memory. I removed it from the command line to do it separately later and restarted the process.
A little after 1 hour, when I restarted where I was, something happened that I had never seen before, the process simply stopped responding and froze. I looked at the chain task and there was a deadlock, but as we already know the process was killed. I restarted again and in 1 hour we will see what happens.
I also noticed that a 24GB java_pid14312.hprof file was generated and there were a few others that I hadn't noticed before.


I was here thinking....
In cases where we have several images to index (in my case more than 35), instead of using several -d on the command line, wouldn't it be more appropriate to process them one by one and use the --append parameter?

**Reply by lfcnassif** [reply]
Processing UI shopping to respond or freezing usually happens before an OutOfMemoryError occurs, it does not mean a dead lock. If your evidences are unrelated (from different persons or computers), using --append is a good approach. Just removing the evidence causing errors from the command line can leave it in an incomplete state into the case, you should remove it from the case using the -remove command line option.

**Reply by lfcnassif** [reply]
And if you can share one of those heap dumps (*.hprof files) I can try to analyze it when I have some available time to try to discover the OOM cause.

**Reply by paulobreim** [reply]
Here is a https://drive.google.com/file/d/1HdUx9NwP3Ur8E5JKmwLnXf_3hqrVUze_/view?usp=sharing
8GB zipped

tks

**Reply by lfcnassif** [reply]
Thanks, I just downloaded. It is being parsed by my Eclipse + Memory Analyzer plugin. But since this heap dump is huge (~34GB), it should take several minutes or a few hours to be parsed. Smaller heap dumps, when created/available, are easier to analyze...

**Reply by lfcnassif** [reply]
Here is a summary of the memory analysis:
![image](https://github.com/sepinf-inc/IPED/assets/7276994/bdeb08c5-b7f2-4011-9ce3-264aa85ce5b7)

65 ZipFile objects are using ~13GB of memory. Looking into them, they come from UFDR decoder related classes. There is an open enhancement to decrease memory usage of UFDR decoder to about half the memory it uses today: #455 

For now, your option is to split your processing to ingest less UFDR evidences at the same time, either using --append or splitting your UFDR evidences across a few cases.

**Reply by paulobreim** [reply]
excellent job.
So far I have 4 ufdr that gave this error.
I removed it from the command line and restarted it and when I'm done and there are no more errors, I'll do these 4 one by one.
My question now is, considering that they had already started, wouldn't it be better to continue processing with --continue?
Or would it be better to -remove and then append?

**Reply by lfcnassif** [reply]
I have never tested that, but I think it is possible to use the same output (-o) used for the whole big case with all evidences and to use just one -d and --continue to resume and finish one evidence at a time.

**Reply by paulobreim** [reply]
New curiosities.
As I didn't know which process was timeout, I started putting it 2 by 2 to find out.
When there was a memory error and I restarted, around 40 files in a single command line, it took around 1:10 to get to the point it was at.
By placing just 2 ufdr on the command line, and I got the first 2 that were theoretically complete, it took 80280 seconds to finish, (22 hours and 13 minutes)!
Now I placed a file alone, and it quickly processed and completed the tasks. But he spends about hours on the "filtering key words" activity, which should take 1 second because there are no words to filter.
Looking at the task manager, the CPU usage is very low, but the output SSD activity is constant.
![Screenshot_20231003_000746](https://github.com/sepinf-inc/IPED/assets/381594/3323361c-3768-4b0d-a4ca-d18a72d43bbb).

These things are interesting to study, unfortunately we have little time for that.


@lfcnassif 
Continuing processing, all files remain in this situation for a long time.
He's probably not in that activity, but I don't know what he's doing.
Considering that it has already processed up to the last file, can I kill the process to go to the next one?

**Reply by paulobreim** [reply]
Progress information:
It's still processing one by one with --continue. It's taking time because after processing the last file, which goes very quickly, it spends more than 1 hour in the situation I mentioned previously of "filtering key words" with low CPU consumption but messing a lot with the output SSD.

**Reply by lfcnassif** [reply]
> it spends more than 1 hour in the situation I mentioned previously of "filtering key words" with low CPU consumption but messing a lot with the output SSD.

Probably it has nothing to to with "keywords", it should be executing another task that hasn't updated the message. I would need a thread dump (jstack -l pid) at this moment to check what is being executed and if it is possible to optimize it.

**Reply by paulobreim** [reply]
tks @lfcnassif , I will try install jstack and get this information.

**Reply by paulobreim** [reply]
Hi @lfcnassif 

There are 2 java.exe.  Here is 2 jstack
[jstack.zip](https://github.com/sepinf-inc/IPED/files/12820584/jstack.zip)

Tks

**Reply by lfcnassif** [reply]
> There are 2 java.exe

Yes, one is the application loader, the other is the main application itself, thanks.

> "main" #1 prio=5 os_prio=0 cpu=13477093.75ms elapsed=14894.02s tid=0x0000000002d7f800 nid=0x16bc runnable  [0x0000000000bee000]
   java.lang.Thread.State: RUNNABLE
	at java.io.FileOutputStream.writeBytes(java.base@11.0.13/Native Method)
	at java.io.FileOutputStream.write(java.base@11.0.13/Unknown Source)
	at java.io.BufferedOutputStream.flushBuffer(java.base@11.0.13/Unknown Source)
	at java.io.BufferedOutputStream.write(java.base@11.0.13/Unknown Source)
	- locked <0x0000000745e7b5d8> (a java.io.BufferedOutputStream)
	at java.io.ObjectOutputStream$BlockDataOutputStream.drain(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream$BlockDataOutputStream.write(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeArray(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeObject0(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeObject(java.base@11.0.13/Unknown Source)
	at java.util.ArrayList.writeObject(java.base@11.0.13/Unknown Source)
	at jdk.internal.reflect.GeneratedMethodAccessor566.invoke(java.base@11.0.13/Unknown Source)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@11.0.13/Unknown Source)
	at java.lang.reflect.Method.invoke(java.base@11.0.13/Unknown Source)
	at java.io.ObjectStreamClass.invokeWriteObject(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeSerialData(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeOrdinaryObject(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeObject0(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.defaultWriteFields(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeSerialData(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeOrdinaryObject(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeObject0(java.base@11.0.13/Unknown Source)
	at java.io.ObjectOutputStream.writeObject(java.base@11.0.13/Unknown Source)
	at iped.engine.util.Util.writeObject(Util.java:287)
	at iped.engine.data.Bookmarks.saveState(Bookmarks.java:436)
	- locked <0x000000013daa3de8> (a iped.engine.data.Bookmarks)
	at iped.engine.data.Bookmarks.saveState(Bookmarks.java:417)
	at iped.engine.task.P2PBookmarker.createBookmarksForSharedFiles(P2PBookmarker.java:132)
	at iped.engine.core.Manager.process(Manager.java:296)
	at iped.app.processing.Main.startManager(Main.java:178)
	at iped.app.processing.Main.execute(Main.java:240)
	at iped.app.processing.Main.main(Main.java:308)

Seems it is taking too long to create bookmarks for files/medias shared by communication apps.

**Reply by paulobreim** [reply]
hmmm is that bookmark that says: probably shared on whatsapp?
It really makes sense.

**Reply by lfcnassif** [reply]
Probably, can be from another chat app too.

**Reply by paulobreim** [reply]
I finally finished processing.
there were 38 cell phones/tablets and 18 notebooks (but only a few folders of them)
Total of just over 50 million records.

Those 4 cell phones that gave OOM, did not give any problems when processed alone.

The only problem I think could be thought of is the last step when it takes a long time to create bookmarks for files/media shared by communication applications.
This happens due to the number of phones and reprocessing them one by one.
Although the default situation is to generate quickly, I think you could put IPEDConfig.txt to generate or not, and by default it would be true.

I intend to redo this entire test in a few days, in fact I already did the last reprocessing in the last version.

Thanks to you both.

**Comment by wladimirleite** [comment]
I created a separated issue (https://github.com/sepinf-inc/IPED/issues/1909) about the Vosk audio transcription slow down, as there are other topics in this discussion.

**Comment by paulobreim** [comment]
New interesting information.
After restarting, it didn't freeze but it gave me an out of memory and then I did the following:
enableExternalParsing = true
externalParsingMaxMem = 1024M
iped -Xms30G -Xmx30G --portable --continue.....
The .ufdr that gave the error before went through without any problems.
I also had the impression that it processed faster.
Below is the screen of processes per task, where half of them have CPU usage.
![Screenshot_20230930_223522](https://github.com/sepinf-inc/IPED/assets/381594/3471f453-41d3-458e-9219-73110bb4486a)

After some time, the next ufdr gave an out of memory with some messages in the console, and also some processes appearing in the task manager, which I believe should not be there. Look how interesting.
![Screenshot_20230930_223722](https://github.com/sepinf-inc/IPED/assets/381594/06aecc96-3e1e-48ba-9321-dd18cb295c83)
![Screenshot_20230930_223744](https://github.com/sepinf-inc/IPED/assets/381594/cda288e5-5c0e-4d77-88fa-38d3a1edfa70)

Can this fix #1910 be used now? If yes, how do I do this?
And what is your opinion on using --append parameter?

tks
Paulo

**Reply by wladimirleite** [reply]
Not a recommendation by any means, but what I do in my own cases is first try to process all evidences in a single command. 
If there is any trouble, I add one by one using --append option. And  --continue if there is still issues.

**Reply by lfcnassif** [reply]
Those messages on the Console are related to an imminent OOM situation. In the Actions section (next to Discussions section) you can find the last snapshot with the VOSK slowdown fix. It's fine to use --append, but if the evidences (disks) came from the same computer (or cellphone), --append can produce a little different results with a few modules depending on the order of evidences in the command line. For example, about WhatsApp processing, if the message database is in the first processed evidence (a cellphone) and if media files attached to chats are in the second processed evidence (a SD card for example) using --append, when the database is processed firstly, the medias in the second evidence wouldn't be available/found, so not linked to chats. Processing the SD card first then appending the main evidence would be a better processing order in this situation.

**Reply by paulobreim** [reply]
Well, it got to the point where it's giving OOM, but I don't know in which file.
So I'm splitting it one by one and running with --continue, since everyone has had most of it processed.

**Comment by lfcnassif** [comment]
> I tried running 4.2-snapshot and although OCR is disabled, the log is showing this:
2023-10-01 16:16:43 [INFO] [app.processing.Main] Indexador e Processador de Evidências Digitais 4.2-snapshot
2023-10-01 16:16:43 [INFO] [engine.config.Configuration] Loading configuration from C:\iped-4.2-snapshot\iped-4.2-snapshot
Warning: Nashorn engine is planned to be removed from a future JDK release
ModuleNotFoundError: No module named 'numpy'
ModuleNotFoundError: No module named 'numpy'
ModuleNotFoundError: No module named 'numpy'
java.lang.RuntimeException: OCRConfig must be loaded before ParsingTaskConfig
at iped.engine.config.ParsingTaskConfig.processProperties(ParsingTaskConfig.java:65)
at iped.engine.config.AbstractTaskPropertiesConfig.processTaskConfig(AbstractTaskPropertiesConfig.java:29)
at iped.engine.config.AbstractTaskConfig.processConfig(AbstractTaskConfig.java:42)
at iped.configuration.Configurable.processConfigs(Configurable.java:40)
at iped.engine.config.ConfigurationManager.loadConfig(ConfigurationManager.java:63)
at iped.engine.config.ConfigurationManager.loadConfigs(ConfigurationManager.java:56)
at iped.engine.config.Configuration.loadConfigurables(Configuration.java:218)
at iped.app.processing.Main.main(Main.java:304)

@paulobreim just to let you know I just fixed this in 4.2-snapshot. It was being caused by conflicts between config files of different IPED versions.

**Reply by paulobreim** [reply]
Ahh thanks for the information.
Taking advantage, let me tell you something that I'm not sure would be some kind of adjustment to be made.
After I started processing file by file, in all cases, at the end of the process, the program took a reasonable amount of time in the task of checking messages that may have been shared. As I was testing a case with more than 40 images and needed to process some more, I decided to remove images that had already been finished from the folder. However, at a certain point, during this task, IPED complained that a file was missing, which was precisely the one I had already processed. This led me to conclude that the task of checking whether a message has been shared performs this check on all files rather than the single file it was processing. This would explain why this time is increasing more and more.
I don't know if this is really how it does it, but in my understanding, sharing would only have to be checked on the file that is being processed.
Does this make sense to you?

**Reply by lfcnassif** [reply]
Yes, it makes sense and the P2PBookmarker class checks for shared files in all processed evidences indeed, not just into the one being processed. This usually is very fast, but in your case it is being a bottle neck. It is possible to optimize it in the future, indeed.

**Reply by paulobreim** [reply]
I understand. I thought that in the WhatsApp database, it was recorded when a file was shared, which would make it much easier.
Thank you for the explanation.
